### PR TITLE
15.0 emoji versions

### PR DESCRIPTION
**Pull request by macchiati**
This should work now. You have to supply an environment variable to generate the beta. Added more documentation to the top of Emoji.java.

### CODE REVIEW

File: unicodetools/src/main/java/org/unicode/tools/emoji/Emoji.java [hunk 1]
```diff
@@ -40,8 +40,49 @@
 import com.ibm.icu.util.ULocale;
 import com.ibm.icu.util.VersionInfo;
 
+/**
+ * You need a command line variable to generate either the beta version (not yet released) or the abbreviated version (see below).
+ * <p>Example: -Demoji-beta
```
Comment (markusicu) [inline unicodetools/src/main/java/org/unicode/tools/emoji/Emoji.java:45]: The emoji-beta option is already mentioned on https://github.com/unicode-org/unicodetools/blob/main/docs/emoji/index.md
Wouldn't it be better to update that and point from here to there?

Note that there are several docs pages under emoji/ : https://github.com/unicode-org/unicodetools/tree/main/docs/emoji

File: unicodetools/src/main/java/org/unicode/tools/emoji/Emoji.java [hunk 2]
```diff
@@ -40,8 +40,49 @@
...
+	…
+</pre>
+*/
 public class Emoji {
 
+	/**
+	 * The following is used to generate an abbreviated version of the charts, where only a few rows are produced,
+	 * and all images are replaced by a colored square (small data size).
+	 * This version can be used to do link-checks.
+	 */
```
Comment (markusicu) [inline unicodetools/src/main/java/org/unicode/tools/emoji/Emoji.java:85]: Please set your editor to write spaces, not tabs, and fix this comment.

**Comment by nedley** [comment]
As I mentioned yesterday, these changes seem like a subset of what I had done on my ned/emoji_15_draft branch so I have to ask: in the generated emoji-data.txt, are you seeing `E15.0`? and does emoji-test.txt contain 15.0 emoji?

**Comment by macchiati** [comment]
Sorry for being so uncommunicative; I'm overbooked right now. Will have a
chance to look at this today. I think there is a switch somewhere to
generate the beta data files; it might be key'ed off of the status of the
candidates (eg provisional v draft) in the candidateData file, but I have
to check. Will have some time later today.

**Comment by macchiati** [comment]
Made modifications to get the files to build. Updated the emoji/index.md. Still need to build emoji-test.txt, but wanted to get these in first. There is a related PR in the emoji repo.

File: unicodetools/src/test/java/org/unicode/propstest/CheckPropertyStability.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
+package org.unicode.propstest;
```
Comment (nedley) [inline unicodetools/src/test/java/org/unicode/propstest/CheckPropertyStability.java:1]: Is this file required for the E15.0 data?

File: unicodetools/src/main/java/org/unicode/tools/emoji/GenerateEmojiData.java [hunk 4]
```diff
@@ -134,7 +134,7 @@ public static ZwjType getType(String s) {
     }
 
     public static <T> void printData() throws IOException {
-        printData(EmojiDataSourceCombined.EMOJI_DATA);
+        printData(EmojiDataSourceCombined.EMOJI_DATA_BETA);
```
Comment (nedley) [inline unicodetools/src/main/java/org/unicode/tools/emoji/GenerateEmojiData.java:137]: I’m assuming this was the missing piece?

File: unicodetools/data/emoji/15.0/emoji-sequences.txt [hunk 5]
```diff
@@ -0,0 +1,1489 @@
...
+1F6BF         ; Basic_Emoji                  ; shower                                                         # E1.0   [1] (🚿)
+1F6C0         ; Basic_Emoji                  ; person taking bath                                             # E0.6   [1] (🛀)
+1F6C1..1F6C5  ; Basic_Emoji                  ; bathtub                                                        # E1.0   [5] (🛁..🛅)
+1F6CC         ; Basic_Emoji                  ; person in bed                                                  # E1.0   [1] (🛌)
+1F6D0         ; Basic_Emoji                  ; place of worship                                               # E1.0   [1] (🛐)
+1F6D1..1F6D2  ; Basic_Emoji                  ; stop sign                                                      # E3.0   [2] (🛑..🛒)
+1F6D5         ; Basic_Emoji                  ; hindu temple                                                   # E12.0  [1] (🛕)
+1F6D6..1F6D7  ; Basic_Emoji                  ; hut                                                            # E13.0  [2] (🛖..🛗)
+1F6DC         ; Basic_Emoji                  ; wireless                                                       # E0.0   [1] (🛜)
```
Comment (nedley) [inline unicodetools/data/emoji/15.0/emoji-sequences.txt:241]: This is still showing `E0.0`.

File: unicodetools/data/emoji/15.0/emoji-data.txt [hunk 6]
```diff
@@ -0,0 +1,1318 @@
...
+303D          ; Emoji                # E0.6   [1] (〽️)       part alternation mark
+3297          ; Emoji                # E0.6   [1] (㊗️)       Japanese “congratulations” button
+3299          ; Emoji                # E0.6   [1] (㊙️)       Japanese “secret” button
+1F004         ; Emoji                # E0.6   [1] (🀄)       mahjong red dragon
+1F0CF         ; Emoji                # E0.6   [1] (🃏)       joker
+1F170..1F171  ; Emoji                # E0.6   [2] (🅰️..🅱️)    A button (blood type)..B button (blood type)
+1F17E..1F17F  ; Emoji                # E0.6   [2] (🅾️..🅿️)    O button (blood type)..P button
+1F18E         ; Emoji                # E0.6   [1] (🆎)       AB button (blood type)
+1F191..1F19A  ; Emoji                # E0.6  [10] (🆑..🆚)    CL button..VS button
+1F1E6..1F1FF  ; Emoji                # E0.0  [26] (🇦..🇿)    regional indicator symbol letter a..regional indicator symbol letter z
```
Comment (nedley) [inline unicodetools/data/emoji/15.0/emoji-data.txt:145]: Should be `E15.0`.

**Review by nedley — COMMENTED** [review]
Thanks for taking the time to flesh this out. Having the 15.0 additions show up as `E0.0` is a problem for the beta but we could fix those by hand if time is running low.

File: unicodetools/src/main/java/org/unicode/tools/emoji/Emoji.java [hunk 7]
```diff
@@ -78,6 +98,7 @@ public class Emoji {
     public static final VersionInfo VERSION0_5 = VersionInfo.getInstance(0,5,2);
 
     // ALSO fix VersionToAge.java!
```
Comment (nedley) [inline unicodetools/src/main/java/org/unicode/tools/emoji/Emoji.java:100]: Is this comment still relevant?

**Comment by nedley** [comment]
@macchiati I will approve your current PRs to not hold things up but at this point we still have issues that are probably blockers for the beta release: charts aside, they are the wrong ages and an incomplete emoji-test.txt. Given your limited availability next week, my proposal is that I make the necessary fixes by hand before trying to debug the tooling but if you know where the problem lies then I will obviously defer to you.

File: unicodetools/data/emoji/15.0/emoji-data.txt (hunk 6 above)
Comment (nedley) [reply unicodetools/data/emoji/15.0/emoji-data.txt:145]: This comment is on the wrong line, sorry: it should refer to a 15.0 emoji as with emoji-sequences.txt.

**Comment by nedley** [comment]
After scrutinizing the updated comments in the code, it sounds like the emoji-test.txt will be fixed after copying the generated files and re-running the tool?

**Comment by macchiati** [comment]
Yes, the E0.0 and missing values in the test are fixed by running the tool again. There is one extra step that I now document in the index.md. See how it looks now.

Haven't tracked down the X issue in the charts yet, but I think this part should be good enough to check in (after checking, of course).

And will fix the X later today.

**Review by nedley — APPROVED** [review]

**Comment by nedley** [comment]
https://github.com/unicode-org/unicodetools/pull/241#discussion_r869473259

**Comment by markusicu** [comment]
I have sent PR #248 for review to move the emoji files to where they belong, sync the duplicated files, and update the readmes.
